browser/ephemeral_storage/ephemeral_storage_qa_browsertest.cc

```diff
@@ -0,0 +1,977 @@
+        prefs::kCookieControlsMode,
+        static_cast<int>(
+            setting ? content_settings::CookieControlsMode::kBlockThirdParty
+                    : content_settings::CookieControlsMode::kOff));
+  }
+
+  void SetCookiePref(ContentSetting setting) {
+    browser()->profile()->GetPrefs()->SetInteger(
+        "profile.default_content_setting_values.cookies", setting);
+  }
+
+  // Starts the JS test code on the QA page to populate storage values so that
+  // they can be read back later in other contexts.
+  void ClickStartTest(content::WebContents* contents) {
+    content::DOMMessageQueue queue(contents);
+    ExecuteScriptAsync(contents,
+                       "document.getElementById('button-start-test').click()");
+    std::string message;
+    while (message != "\"button operation completed\"") {
+      ASSERT_TRUE(queue.WaitForMessage(&message));
+    }
+  }
+
+  // Prepares the test page for validation of test results by reading storage
+  // and writing to the 2D results matrix.
+  void ClickReadValues(content::WebContents* contents) {
+    content::DOMMessageQueue queue(contents);
+    ExecuteScriptAsync(contents,
+                       "document.getElementById('button-read-values').click()");
+    std::string message;
+    while (message != "\"button operation completed\"") {
+      ASSERT_TRUE(queue.WaitForMessage(&message));
+    }
+  }
+
+  // Performs a navigation in the current session to the other-origin page by
+  // clicking the link with the corresponding href attribute.
+  void NavigateOtherOrigin(content::WebContents* contents) {
+    {
+      TabActivationWaiter tab_activation_waiter(browser()->tab_strip_model());
+      ExecuteScriptAsync(
+          contents,
+          "document.querySelector('.other-origin.ephem-storage-test').click()");
+      tab_activation_waiter.WaitForActiveTabChange();
+    }
+    content::TestNavigationObserver navigation_observer(
+        browser()->tab_strip_model()->GetActiveWebContents());
+    navigation_observer.Wait();
+    ASSERT_TRUE(navigation_observer.last_navigation_succeeded());
+  }
+
+  // Performs a navigation in the current session to the same-origin page by
+  // clicking the link with the corresponding href attribute.
+  void NavigateSameOrigin(content::WebContents* contents) {
+    {
+      TabActivationWaiter tab_activation_waiter(browser()->tab_strip_model());
+      ExecuteScriptAsync(
+          contents,
+          "document.querySelector('.this-origin.ephem-storage-test').click()");
+      tab_activation_waiter.WaitForActiveTabChange();
+    }
+    content::TestNavigationObserver navigation_observer(
+        browser()->tab_strip_model()->GetActiveWebContents());
+    navigation_observer.Wait();
+    ASSERT_TRUE(navigation_observer.last_navigation_succeeded());
+  }
+
+  // Checks that the 2D storage results matrix on the test page matches the
+  // expected values
+  void CheckStorageResults(content::WebContents* contents,
```

Comment:
it seems a bit strange to check the results matrix instead of making actual cookie/storage values using normal testing. Also there's no way in the page to check network cookies. We want to convert the matrix used for QA into actual test assertions both for simplicity and also for maintainability of the test itself